File: 프로그래머스_모의고사/2차_문제집/송기훈/2번.py
# 10점
# def solution(topping):
#     answer = 0

#     for i in range(len(topping)):
#         a = topping[:i]
#         b = topping[i:]

#         setA = set(a)
#         setB = set(b)

#         if len(setA) == len(setB):
#             answer += 1

#     return answer


def solution(topping):
    answer = 0

    a = dict()
    b = dict()

    for t in topping:
        if t not in b:
            b[t] = 1
        else:
            b[t] += 1

    for i in range(len(topping)):
        temp = topping[i]
        b[temp] -= 1

        if temp not in a:
            a[temp] = 1
        else:
            a[temp] += 1

        if b[temp] == 0:
            del b[temp]
        if len(a) == len(b):
            answer += 1

    return answer

File: 프로그래머스_모의고사/2차_문제집/이아름/2번.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    map<int,int> total;
    for(int n : topping){
        total[n]++;
    }
    map<int,int> right;
    for(int n : topping){
        total[n]--;
        right[n]++;
        if(total[n] == 0){
            total.erase(n);
        }
        if(total.size() == right.size()){
            answer++;
        }
    }

    return answer;
}
